### PR TITLE
Fix [TempData] for PageModels

### LINKED ISSUE #6074: [TempData] Properties don't work on PageModel classes

### PR DESCRIPTION
**Pull request by ryanbrandenburg**
Fixes #6074.

### CODE REVIEW

**Comment by DamianEdwards** [comment]
It's happening

File: src/Microsoft.AspNetCore.Mvc.ViewFeatures/Internal/SaveTempDataPropertyFilter/ControllerSaveTempDataPropertyFilterFactory.cs [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+using System.Collections.Generic;
+using Microsoft.AspNetCore.Mvc.Filters;
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.Internal;
+
+namespace Microsoft.AspNetCore.Mvc.ViewFeatures.Internal
+{
+    public class ControllerSaveTempDataPropertyFilterFactory : IFilterFactory
+    {
+        // Cannot be public as <c>PropertyHelper</c> is an internal shared source type
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.ViewFeatures/Internal/SaveTempDataPropertyFilter/ControllerSaveTempDataPropertyFilterFactory.cs]: Remove the comment

File: src/Microsoft.AspNetCore.Mvc.ViewFeatures/Internal/SaveTempDataPropertyFilter/PageSaveTempDataPropertyFilter.cs [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+using Microsoft.AspNetCore.Http;
+
+namespace Microsoft.AspNetCore.Mvc.ViewFeatures.Internal
+{
+    public class PageSaveTempDataPropertyFilter : SaveTempDataPropertyFilterBase
+    {
+        public PageSaveTempDataPropertyFilter(ITempDataDictionaryFactory factory)
+            : base(factory)
+        {
+            OriginalValues = new Dictionary<PropertyInfo, object>();
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.ViewFeatures/Internal/SaveTempDataPropertyFilter/PageSaveTempDataPropertyFilter.cs]: Should this just be initialized as part of the base class?

Comment (ryanbrandenburg) [inline src/Microsoft.AspNetCore.Mvc.ViewFeatures/Internal/SaveTempDataPropertyFilter/PageSaveTempDataPropertyFilter.cs]: Yeah, right now `OriginalValues` is initialized sort of piecemeal across a couple differnt methods depending on which you run first, I'll try initializing it in the base constructor and see if anything breaks.

File: src/Microsoft.AspNetCore.Mvc.ViewFeatures/Internal/SaveTempDataPropertyFilter/SaveTempDataPropertyFilterBase.cs [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+
+        /// <summary>
+        /// Puts the modified values of <see cref="Subject"/> into <paramref name="tempData"/>.
+        /// </summary>
+        /// <param name="tempData">The <see cref="ITempDataDictionary"/> to be updated.</param>
+        public void OnTempDataSaving(ITempDataDictionary tempData)
+        {
+            if (Subject != null && OriginalValues != null)
+            {
+                foreach (var kvp in OriginalValues)
```
Comment (ryanbrandenburg) [inline src/Microsoft.AspNetCore.Mvc.ViewFeatures/Internal/SaveTempDataPropertyFilter/SaveTempDataPropertyFilterBase.cs]: Do we need to keep track of the OriginalValues at all? Couldn't we just iterate over our TempDataProperties and apply them to tempData without knowing or caring if it changed? Is the cost of setting a tempData value higher than I think?

**Comment by ryanbrandenburg** [comment]
:up::date:

File: src/Microsoft.AspNetCore.Mvc.ViewFeatures/Internal/SaveTempDataPropertyFilter/ControllerSaveTempDataPropertyFilterFactory.cs [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+            {
+                throw new ArgumentNullException(nameof(serviceProvider));
+            }
+
+            var service = serviceProvider.GetRequiredService<ControllerSaveTempDataPropertyFilter>();
+            service.TempDataProperties = TempDataProperties;
+            return service;
+        }
+
+        public static IList<TempDataProperty> GetTempDataPropertyHelpers(Type modelType)
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.ViewFeatures/Internal/SaveTempDataPropertyFilter/ControllerSaveTempDataPropertyFilterFactory.cs]: Does this need to be public? Would `protected` work here?

Just call the parameter `type` - it isn't necessarily a model type.

Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.ViewFeatures/Internal/SaveTempDataPropertyFilter/ControllerSaveTempDataPropertyFilterFactory.cs]: This should probably be on `SaveTempDataPropertyFilterBase` if it's going to be public.

File: src/Microsoft.AspNetCore.Mvc.ViewFeatures/Internal/SaveTempDataPropertyFilter/PageSaveTempDataPropertyFilter.cs [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+        /// <param name="modelType">The <see cref="Type"/> to create <see cref="TempDataProperty"/>'s for.</param>
+        public void SetTempDataProperties(Type modelType)
+        {
+            if (FilterFactory == null)
+            {
+                throw new ArgumentNullException(nameof(FilterFactory));
+            }
+
+            TempDataProperties = FilterFactory.GetTempDataProperties(modelType);
+        }
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.ViewFeatures/Internal/SaveTempDataPropertyFilter/PageSaveTempDataPropertyFilter.cs]: One thing that you could do here is you could make `Subject` virtual, and you could do this work when the property is set.

File: src/Microsoft.AspNetCore.Mvc.ViewFeatures/Internal/SaveTempDataPropertyFilter/PageSaveTempDataPropertyFilter.cs [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+        /// <summary>
+        /// Creates and applies the set of <see cref="TempDataProperty"/>'s for the given <paramref name="modelType"/>.
+        /// </summary>
+        /// <param name="modelType">The <see cref="Type"/> to create <see cref="TempDataProperty"/>'s for.</param>
+        public void SetTempDataProperties(Type modelType)
+        {
+            if (FilterFactory == null)
+            {
+                throw new ArgumentNullException(nameof(FilterFactory));
+            }
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.ViewFeatures/Internal/SaveTempDataPropertyFilter/PageSaveTempDataPropertyFilter.cs]: This isn't the right null check

File: src/Microsoft.AspNetCore.Mvc.ViewFeatures/Internal/SaveTempDataPropertyFilter/PageSaveTempDataPropertyFilter.cs [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+        /// Applies values from TempData from <paramref name="httpContext"/> to the
+        /// <see cref="SaveTempDataPropertyFilterBase.Subject"/>.
+        /// </summary>
+        /// <param name="httpContext">The <see cref="HttpContext"/> used to find TempData.</param>
+        public void ApplyTempDataChanges(HttpContext httpContext)
+        {
+            if (Subject == null)
+            {
+                throw new ArgumentNullException(nameof(Subject));
+            }
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.ViewFeatures/Internal/SaveTempDataPropertyFilter/PageSaveTempDataPropertyFilter.cs]: This isn't right.

File: src/Microsoft.AspNetCore.Mvc.ViewFeatures/Internal/SaveTempDataPropertyFilter/PageSaveTempDataPropertyFilterFactory.cs [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+                throw new ArgumentNullException(nameof(serviceProvider));
+            }
+
+            var service = serviceProvider.GetRequiredService<PageSaveTempDataPropertyFilter>();
+            service.FilterFactory = this;
+
+            return service;
+        }
+
+        public IList<TempDataProperty> GetTempDataProperties(Type modelType)
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.ViewFeatures/Internal/SaveTempDataPropertyFilter/PageSaveTempDataPropertyFilterFactory.cs]: Make sure to leave some comments behind for future generations about why this is so complex.

File: src/Microsoft.AspNetCore.Mvc.ViewFeatures/Internal/SaveTempDataPropertyFilter/SaveTempDataPropertyFilterBase.cs [hunk 9]
```diff
@@ -0,0 +1,87 @@
...
+
+        /// <summary>
+        /// The <see cref="object"/> which has the temp data properties.
+        /// </summary>
+        public object Subject { get; set; }
+
+        /// <summary>
+        /// Tracks the values which originally existed in temp data.
+        /// </summary>
+        public IDictionary<PropertyInfo, object> OriginalValues { get; set; } = new Dictionary<PropertyInfo, object>();
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.ViewFeatures/Internal/SaveTempDataPropertyFilter/SaveTempDataPropertyFilterBase.cs]: You can make this get-only

File: src/Microsoft.AspNetCore.Mvc.ViewFeatures/Internal/SaveTempDataPropertyFilter/SaveTempDataPropertyFilterBase.cs (hunk 3 above)
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.ViewFeatures/Internal/SaveTempDataPropertyFilter/SaveTempDataPropertyFilterBase.cs]: We want to use `OriginalValues`. TempData cares when you set values.

File: src/Microsoft.AspNetCore.Mvc.ViewFeatures/Internal/SaveTempDataPropertyFilter/TempDataProperty.cs [hunk 10]
```diff
@@ -0,0 +1,34 @@
...
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Reflection;
+
+namespace Microsoft.AspNetCore.Mvc.ViewFeatures.Internal
+{
+    public struct TempDataProperty
+    {
+        public PropertyInfo PropertyInfo { get; private set; }
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.ViewFeatures/Internal/SaveTempDataPropertyFilter/TempDataProperty.cs]: No need for the `private set`.

Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.ViewFeatures/Internal/SaveTempDataPropertyFilter/TempDataProperty.cs]: Properties after constructor

File: test/Microsoft.AspNetCore.Mvc.FunctionalTests/RazorPagesTest.cs [hunk 11]
```diff
@@ -21,6 +21,17 @@ public RazorPagesTest(MvcTestFixture<RazorPagesWebSite.Startup> fixture)
...
 
         [Fact]
+        public async Task RazorPage_TempDataFilters_DontSurviveBetweenPages()
+        {
+            var response = await Client.GetAsync("http://localhost/TempData/TempDataPageModelProperty");
+            Assert.Equal(HttpStatusCode.OK, response.StatusCode);
+
+            //If the filter from above was re-used in the below request the request would fail due to type cast.
+            response = await Client.GetAsync("http://localhost/HandlerTestPage");
+            Assert.Equal(HttpStatusCode.OK, response.StatusCode);
```
Comment (rynowak) [inline test/Microsoft.AspNetCore.Mvc.FunctionalTests/RazorPagesTest.cs]: This is more suited to a unit test.

Comment (rynowak) [inline test/Microsoft.AspNetCore.Mvc.FunctionalTests/RazorPagesTest.cs]: In fact I'm not really sure what this is testing.

File: test/Microsoft.AspNetCore.Mvc.ViewFeatures.Test/Internal/SaveTempDataFilter/SaveTempDataPropertyFilterHelper.cs [hunk 12]
```diff
@@ -0,0 +1,28 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System.Collections.Generic;
+using System.Reflection;
+
+namespace Microsoft.AspNetCore.Mvc.ViewFeatures.Internal
+{
+    public static class SaveTempDataPropertyFilterHelper
```
Comment (rynowak) [inline test/Microsoft.AspNetCore.Mvc.ViewFeatures.Test/Internal/SaveTempDataFilter/SaveTempDataPropertyFilterHelper.cs]: No.

Comment (rynowak) [inline test/Microsoft.AspNetCore.Mvc.ViewFeatures.Test/Internal/SaveTempDataFilter/SaveTempDataPropertyFilterHelper.cs]: Make a base class for the tests or just call into the product code to create these.

File: test/Microsoft.AspNetCore.Mvc.FunctionalTests/RazorPagesTest.cs (hunk 11 above)
Comment (ryanbrandenburg) [inline test/Microsoft.AspNetCore.Mvc.FunctionalTests/RazorPagesTest.cs]: It's testing that Factories (and therefore TempDataProperties) are per-page and don't "leak". This test failed before I added the new `ConfigureFilter` overload and used that. I tried to describe that in the comment above, but I'll take another stab at clarifying.

Comment (ryanbrandenburg) [inline test/Microsoft.AspNetCore.Mvc.FunctionalTests/RazorPagesTest.cs]: Or do we think that's the kind of bug that will only happen while writing it and we don't need the test?

Comment (rynowak) [inline test/Microsoft.AspNetCore.Mvc.FunctionalTests/RazorPagesTest.cs]: Bingo. I wouldn't write a test for code that we don't have.

**Comment by ryanbrandenburg** [comment]
:up::date:

File: src/Microsoft.AspNetCore.Mvc.ViewFeatures/Internal/SaveTempDataPropertyFilter/PageSaveTempDataPropertyFilter.cs [hunk 13]
```diff
@@ -0,0 +1,64 @@
...
+        public PageSaveTempDataPropertyFilter(ITempDataDictionaryFactory factory)
+            : base(factory)
+        {
+        }
+
+        public PageSaveTempDataPropertyFilterFactory FilterFactory { get; set; }
+
+        public override object Subject {
+            get => base.Subject;
+            set {
```
Comment (pranavkm) [inline src/Microsoft.AspNetCore.Mvc.ViewFeatures/Internal/SaveTempDataPropertyFilter/PageSaveTempDataPropertyFilter.cs]: Braces on next line.

File: src/Microsoft.AspNetCore.Mvc.ViewFeatures/Internal/SaveTempDataPropertyFilter/SaveTempDataPropertyFilterBase.cs [hunk 14]
```diff
@@ -0,0 +1,134 @@
...
+                    {
+                        tempData[Prefix + property.Name] = newValue;
+                    }
+                }
+            }
+        }
+
+        public static IList<TempDataProperty> GetTempDataProperties(Type type)
+        {
+            IList<TempDataProperty> results = null;
```
Comment (pranavkm) [inline src/Microsoft.AspNetCore.Mvc.ViewFeatures/Internal/SaveTempDataPropertyFilter/SaveTempDataPropertyFilterBase.cs]: `List<TempDataProperty> results` (no need to declare it as an interface type)

File: test/Microsoft.AspNetCore.Mvc.ViewFeatures.Test/Internal/SaveTempDataFilter/SaveTempDataPropertyFilterTestBase.cs [hunk 15]
```diff
@@ -0,0 +1,28 @@
...
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System.Collections.Generic;
+using System.Reflection;
+
+namespace Microsoft.AspNetCore.Mvc.ViewFeatures.Internal
+{
+    public class SaveTempDataPropertyFilterTestBase
+    {
+        public static IList<TempDataProperty> BuildTempDataProperties<TSubject>()
```
Comment (pranavkm) [inline test/Microsoft.AspNetCore.Mvc.ViewFeatures.Test/Internal/SaveTempDataFilter/SaveTempDataPropertyFilterTestBase.cs]: Could we not have this type? It's used in 4 places and we can construct the `TempDataProperty` manually in a helper method sans reflection

File: test/WebSites/RazorPagesWebSite/TempData/TempDataPageModelProperty.cshtml [hunk 16]
```diff
@@ -0,0 +1,9 @@
+﻿@page
+@model RazorPagesWebSite.TempData.TempDataPageModel
+
+Message: @Model.Message
+@using (Html.BeginForm())
+{
+    @Html.AntiForgeryToken()
+}
+TempData: @TempData["TempDataProperty-Message"]
```
Comment (pranavkm) [inline test/WebSites/RazorPagesWebSite/TempData/TempDataPageModelProperty.cshtml]: Trailing new line

File: test/Microsoft.AspNetCore.Mvc.ViewFeatures.Test/Internal/SaveTempDataFilter/PageSaveTempDataPropertyFilterTest.cs [hunk 17]
```diff
@@ -0,0 +1,221 @@
...
+        }
+
+        [Fact]
+        public void SetSubject_ModifiesFactoryAndFilter()
+        {
+            // Arrange
+            var httpContext = new DefaultHttpContext();
+
+            var tempData = new TempDataDictionary(httpContext, Mock.Of<ITempDataProvider>())
+            {
```
Comment (pranavkm) [inline test/Microsoft.AspNetCore.Mvc.ViewFeatures.Test/Internal/SaveTempDataFilter/PageSaveTempDataPropertyFilterTest.cs]: Remove

File: test/Microsoft.AspNetCore.Mvc.ViewFeatures.Test/Internal/SaveTempDataFilter/SaveTempDataPropertyFilterTestBase.cs (hunk 15 above)
Comment (pranavkm) [inline test/Microsoft.AspNetCore.Mvc.ViewFeatures.Test/Internal/SaveTempDataFilter/SaveTempDataPropertyFilterTestBase.cs]: ^

File: test/Microsoft.AspNetCore.Mvc.ViewFeatures.Test/Internal/SaveTempDataFilter/ControllerSaveTempDataPropertyFilterTest.cs [hunk 18]
```diff
@@ -62,10 +69,17 @@ public void SaveTempDataPropertyFilter_ReadsTempDataFromTempDataDictionary()
...
-            var filter = CreateSaveTempDataPropertyFilter(httpContext, tempData: tempData);
+            var filter = CreateControllerSaveTempDataPropertyFilter(httpContext, tempData: tempData);
             var controller = new TestController();
 
-            filter.PropertyHelpers = BuildPropertyHelpers<TestController>();
+            var controllerType = controller.GetType();
+            var testProp = controllerType.GetProperty(nameof(TestController.Test));
+            var test2Prop = controllerType.GetProperty(nameof(TestController.Test2));
+
+            filter.TempDataProperties = new List<TempDataProperty> {
```
Comment (pranavkm) [inline test/Microsoft.AspNetCore.Mvc.ViewFeatures.Test/Internal/SaveTempDataFilter/ControllerSaveTempDataPropertyFilterTest.cs]: bracing 😢

**Pull request merged**